lib/munster/models/received_webhook.rb

```diff
@@ -1,23 +1,107 @@
+      are getting saved with the request headers intact.
+    EOS
+
     self.implicit_order_column = "created_at"
     self.table_name = "received_webhooks"
 
     include StateMachineEnum
 
     state_machine_enum :status do |s|
       s.permit_transition(:received, :processing)
+      s.permit_transition(:received, :failed_validation)
       s.permit_transition(:processing, :skipped)
       s.permit_transition(:processing, :processed)
       s.permit_transition(:processing, :error)
     end
 
+    # Store the pertinent data from an ActionDispatch::Request into the webhook.
+    # @param [ActionDispatch::Request]
+    def request=(action_dispatch_request)
+      # Filter out all Rack-specific headers such as "rack.input" and the like. We are
+      # only interested in the actual HTTP headers presented by the webserver. Mostly...
+      headers = action_dispatch_request.env.filter_map do |(request_header, header_value)|
+        if request_header.is_a?(String) && request_header.upcase == request_header && header_value.is_a?(String)
+          [request_header, header_value]
+        end
+      end.to_h
+
+      # ...except the path parameters - they do not get parsed from the headers, but instead get set by Journey - the Rails
+      # router - when the ActionDispatch::Request object gets instantiated. They need to be preserved separately in case the Munster
+      # controller gets mounted under a parametrized path - and the path component actually is a parameter that the webhook
+      # handler either needs for validation or for processing
+      headers["action_dispatch.request.path_parameters"] = action_dispatch_request.env.fetch("action_dispatch.request.path_parameters")
+
+      # If the migration hasn't been applied yet, we can't save the headers.
+      if self.class.column_names.include?("request_headers")
+        write_attribute("request_headers", headers)
+      else
+        Rails.logger.warn { "You need to run Munster migrations so that request headers can be persisted with the model. Async validation is not going to work without that column being set." }
+      end
+      write_attribute("body", action_dispatch_request.body.read.force_encoding(Encoding::BINARY))
+    ensure
+      action_dispatch_request.body.rewind
+    end
+
+    # A Munster handler is, in a way, a tiny Rails controller which runs in a background job. To allow this,
+    # we need to provide access not only to the webhook payload (the HTTP request body, usually), but also
+    # to the rest of the HTTP request - such as headers and route params. For example, imagine you use
+    # a system where your multiple tenants (users) may receive webhooks from the same sender. However, you
+    # need to dispatch those webhooks to those particular tenants of your application. Instead of mounting
+    # Munster as an engine under a common route, you mount it like so:
+    #
+    #    post "/incoming-webhooks/:user_id/:service_id" => "munster/receive_webhooks#create"
```

Comment:
Not really. The README is quite thin at the moment, and having comments in the code right now provides for a more refined medium - we can add explanations closer to where people are likely to look. Also, this describes a special case of a special use of the ActionDispatch request object, which is exactly the thing I tend to explain in comments (for example: why isn't a `Rack::Request` enough? because it does not provide `params`, for one)